pkg/schedule/coordinator.go

```diff
@@ -152,72 +149,195 @@ func (c *Coordinator) IsPendingRegion(region uint64) bool {
+
+	ticker := time.NewTicker(c.patrolRegionContext.interval)
 	defer ticker.Stop()
 
+	c.patrolRegionContext.startPatrolRegionWorkers(c)
+
 	log.Info("coordinator starts patrol regions")
-	start := time.Now()
 	var (
 		key     []byte
 		regions []*core.RegionInfo
 	)
 	for {
 		select {
 		case <-ticker.C:
-			// Note: we reset the ticker here to support updating configuration dynamically.
-			ticker.Reset(c.cluster.GetCheckerConfig().GetPatrolRegionInterval())
+			c.updateTickerIfNeeded(ticker)
+			c.updatePatrolWorkersIfNeeded()
+
+			if c.cluster.IsSchedulingHalted() {
+				for len(c.patrolRegionContext.regionChan) > 0 {
+					<-c.patrolRegionContext.regionChan
+				}
+				continue
+			}
+
+			// wait for the regionChan to be drained
+			if len(c.patrolRegionContext.regionChan) > 0 {
+				continue
+			}
+
+			// Check priority regions first.
+			c.checkPriorityRegions()
+			// Check suspect regions first.
+			c.checkSuspectRegions()
+			// Check regions in the waiting list
+			c.checkWaitingRegions()
+
+			key, regions = c.checkRegions(key)
+			if len(regions) == 0 {
+				continue
+			}
+			// Updates the label level isolation statistics.
+			c.cluster.UpdateRegionsLabelLevelStats(regions)
+			// Update metrics if a full scan is done.
+			if len(key) == 0 {
+				c.patrolRegionContext.roundUpdate()
+				c.patrolRegionContext.updateScanLimit(c.cluster)
+			}
+			failpoint.Inject("break-patrol", func() {
+				time.Sleep(100 * time.Millisecond) // ensure the regions are handled by the workers
+				failpoint.Return()
+			})
 		case <-c.ctx.Done():
 			patrolCheckRegionsGauge.Set(0)
-			c.setPatrolRegionsDuration(0)
+			c.patrolRegionContext.setPatrolRegionsDuration(0)
 			log.Info("patrol regions has been stopped")
 			return
 		}
-		if c.cluster.IsSchedulingHalted() {
-			continue
-		}
+	}
+}
 
-		// Check priority regions first.
-		c.checkPriorityRegions()
-		// Check suspect regions first.
-		c.checkSuspectRegions()
-		// Check regions in the waiting list
-		c.checkWaitingRegions()
+func (c *Coordinator) updateTickerIfNeeded(ticker *time.Ticker) {
+	// Note: we reset the ticker here to support updating configuration dynamically.
+	newInterval := c.cluster.GetCheckerConfig().GetPatrolRegionInterval()
+	if c.patrolRegionContext.interval != newInterval {
+		c.patrolRegionContext.interval = newInterval
+		ticker.Reset(newInterval)
+	}
+}
 
-		key, regions = c.checkRegions(key)
-		if len(regions) == 0 {
-			continue
-		}
-		// Updates the label level isolation statistics.
-		c.cluster.UpdateRegionsLabelLevelStats(regions)
-		if len(key) == 0 {
-			dur := time.Since(start)
-			patrolCheckRegionsGauge.Set(dur.Seconds())
-			c.setPatrolRegionsDuration(dur)
-			start = time.Now()
-		}
-		failpoint.Inject("break-patrol", func() {
-			failpoint.Break()
-		})
+func (c *Coordinator) updatePatrolWorkersIfNeeded() {
+	newWorkersCount := c.cluster.GetCheckerConfig().GetPatrolRegionWorkerCount()
+	if c.patrolRegionContext.workerCount != newWorkersCount {
```

Comment:
Can we wrap the read/write operation inside patrolRegionContext so that we can avoid the lock issue.

Reply:
ok, I will add a lock for config and status.